widgets-tileset/index.ts

```diff
@@ -0,0 +1,356 @@
+  if (droppingPercent > EXCESIVE_DROPPING_PERCENT) {
+    droppingWarningBig.classList.remove('hidden');
+    wrappers.forEach(el => el.classList.add('dim'));
+  } else if (droppingPercent > 0) {
+    droppingWarningSmall.classList.remove('hidden');
+  }
+
+  dataSource.widgetSource.extractTileFeatures({ spatialFilter: viewportSpatialFilter })
+
+  histogramWidgetChart.showLoading();
+
+  // configure widgets
+
+  const formula = await dataSource.widgetSource.getFormula({
+    column: '*',
+    operation: 'count'
+  })
+
+  if (formula.value) {
+    formulaWidget.innerHTML = formula.value.toFixed(0);
+  }
+
+  const histogram = await dataSource.widgetSource.getHistogram({
+    column: 'streamOrder',
+    ticks: histogramTicks,
+    operation: 'count'
+  });
+
+  const option = {
+    tooltip: {
+      trigger: 'axis',
+      axisPointer: {
+        type: 'shadow'
+      }
+    },
+    grid: {
+      left: 60,
+      right: 30,
+      top: 20,
+      bottom: 20,
+      width: 'auto',
+      height: 'auto'
+    },
+    xAxis: {
+      type: 'category',
+      data: histogramTicks,
+      // axisLabel: {
+      //   interval: 4 // Show every 5th label
+      // },
+      axisTick: {
+        alignWithLabel: true
+      }
+    },
+    yAxis: {
+      type: 'value',
+      axisLabel: {
+        formatter: (value: number) => Intl.NumberFormat('en-US', { compactDisplay: 'short', notation: 'compact' }).format(value)
+      }
+    },
+    series: [
+      {
+        name: 'Count',
+        type: 'bar',
+        data: histogram,
+        itemStyle: {
+          color: '#3398DB'
+        }
+      }
+    ]
+  };
+
+  // render widgets!
+
+  histogramWidgetChart.setOption(option);
+  histogramWidgetChart.hideLoading();
+}
+
+const colors = [
+  '#f7fcf0',
+  '#e0f3db',
+  '#ccebc5',
+  '#a8ddb5',
+  '#7bccc4',
+  '#4eb3d3',
+  '#2b8cbe',
+  '#08589e',
+].map((hex) => hexToRgb(hex));
+
+function hexToRgb(hex) {
+  const r = parseInt(hex.slice(1, 3), 16);
+  const g = parseInt(hex.slice(3, 5), 16);
+  const b = parseInt(hex.slice(5, 7), 16);
+  return [r, g, b];
+}
+
+// render Layers function
+
+function renderLayers() {
+  // Exit if dataSource is not ready
+  if (!dataSource) {
+    return;
+  }
+
+  // now for the layers
+
+  const layers = [
+    new VectorTileLayer({
+      id: 'places',
+      pickable: true,
+      data: dataSource,
+      opacity: 1,
+      getLineColor: d => {
+        const [r, g, b] = hexToRgb('#d5d5d7')
+        const n = d.properties.streamOrder;
+        const alphaPart = Math.min(n / 10, 1);
+        const alpha = 120 + (128 * alphaPart);
+        return [r, g, b, alpha]
+      },
+      getLineWidth: d => {
+        const n = d.properties.streamOrder;
+        return n * 0.5;
+      },
+      lineWidthUnits: 'pixels',
+      lineWidthMinPixels: 1,
+      onViewportLoad(tiles) {
+        dataSource.widgetSource.loadTiles(tiles)
+        if (!tilesLoaded) {
+          tilesLoaded = true;
+          renderWidgets();
```

Comment:
Very minor and doesn't necessarily need to be changed, but would it be better to call something debounced like `debouncedUpdateSpatialFilter` here? I don't expect the `onViewportLoad` callback to run that often, but it might trigger just before/after a viewstate change, and so we'd end up doing the calculations twice since they don't share a debounce. 

Would matter a bit more for tilesets with many features, or for rasters.

I've done something similar in:

https://github.com/CartoDB/carto-api-client/pull/87

Reply:
Done.
This came from a bug where `onViewStateChange` was being called before `onViewportLoad` was called for the first time, so the app was trying to do an `extractFeatures` before having called `loadTiles`, so I made sure to set a boolean for a loading state there and to call `extractFeatures` also depending on that boolean in `onViewportLoad`